py/server/deephaven/execution_context.py

```diff
@@ -0,0 +1,98 @@
+    evaluate queries submitted through this script session. However, to be able to evaluate a query in a deferred
+    manner, such as in a different thread from the script session thread, an application must explicitly obtain or
+    build a ExecutionContext and either (1) install it on the thread or (2) use it as a context manager to enclose
+    the query in the body of the with statement.
+
+    Note that, a ExecutionContext can be shared among many threads. The most typical use pattern would be to obtain
+    the script session's systemic ExecutionContext and set it on a thread explicitly created by the user or simply
+    use it to wrap a query run in that thread.
+    """
+    j_object_type = _JExecutionContext
+
+    @property
+    def j_object(self) -> jpy.JType:
+        return self.j_exec_ctx
+
+    def __init__(self, j_exec_ctx):
+        self.j_exec_ctx = j_exec_ctx
+
+    def __enter__(self):
+        self._j_safe_closable = self.j_exec_ctx.open()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._j_safe_closable.close()
+
+
+def make_user_exec_ctx(freeze_vars: Union[str, Sequence[str]] = None) -> ExecutionContext:
+    """Makes a new ExecutionContext based off the current thread's ExecutionContext. The optional parameter
+    freeze_vars should be a list of names in the current query scope. If it is provided, the result ExecutionContext
+    will include a new query scope that is made up of only these names together with their current values. Future
+    changes to the values of these names will not be visible in the new ExecutionContext.
+
+    Args:
+        freeze_vars(Union[str, Sequence[str]]): the names of the vars in the current query scope to be frozen in
+            the new one
+
+    Returns:
+        a ExecutionContext
+
+    Raises:
+        DHError
+    """
+    freeze_vars = to_sequence(freeze_vars)
+
+    if not freeze_vars:
+        return ExecutionContext(j_exec_ctx=_JExecutionContext.makeExecutionContext(False))
+    else:
+        try:
+            j_exec_ctx = (_JExecutionContext.newBuilder()
+                          .captureQueryCompiler()
+                          .captureQueryLibrary()
+                          .captureQueryScopeVars(*freeze_vars)
+                          .build())
+            return ExecutionContext(j_exec_ctx=j_exec_ctx)
+        except Exception as e:
+            raise DHError(message="failed to make a new ExecutionContext") from e
+
+
+def get_exec_ctx() -> ExecutionContext:
+    """Returns the current thread's ExecutionContext.
+
+    Returns:
+        a ExecutionContext
+    """
+    return ExecutionContext(j_exec_ctx=_JExecutionContext.getContext())
+
+
+def set_exec_ctx(exec_ctx: ExecutionContext) -> None:
+    """Sets the current thread's ExecutionContext.
+
+    Args:
+        exec_ctx (ExecutionContext): the ExecutionContext to be installed on the current thread
+    """
+    _JExecutionContext.setContext(exec_ctx.j_exec_ctx)
```

Comment:
I don't think we should really support this. We should enforce the usage of `open` and `close`.